From 312c8ddc0a68c560d4598320b2ff570e7c11cde9 Mon Sep 17 00:00:00 2001
From: Kurt Newman <kurt.newman@cpanel.net>
Date: Tue, 13 Jan 2015 18:14:44 -0600
Subject: [PATCH 2/7] Allow scripts to show itself in process list

Case 4546: When this value is set to true, mod_suphp
will execute PHP scripts in a way that displays both
the PHP interpreter and the SCRIPT_FILENAME in the
process list, generated by the ps command. Set this
value to false to hide the SCRIPT_FILENAME and improve
security.

More information can be found at:
 http://documentation.cpanel.net/display/EA/Apache+PHP+Request+Handling
---
 src/Application.cpp   | 3 +++
 src/Configuration.cpp | 7 +++++++
 src/Configuration.hpp | 7 +++++++
 3 files changed, 17 insertions(+)

diff --git a/src/Application.cpp b/src/Application.cpp
index d8e8ff8..672d743 100644
--- a/src/Application.cpp
+++ b/src/Application.cpp
@@ -543,6 +543,9 @@ void suPHP::Application::executeScript(const std::string& scriptFilename,
             std::string interpreterPath = interpreter.substr(4);
             CommandLine cline;
             cline.putArgument(interpreterPath);
+            if ( config.getFullPHPProcessDisplay() ) {
+	        cline.putArgument(scriptFilename);
+            }
             API_Helper::getSystemAPI().execute(interpreterPath, cline, env);
         } else if (mode == TARGETMODE_SELFEXECUTE) {
             CommandLine cline;
diff --git a/src/Configuration.cpp b/src/Configuration.cpp
index fbd0c46..0c63ff5 100644
--- a/src/Configuration.cpp
+++ b/src/Configuration.cpp
@@ -113,6 +113,7 @@ suPHP::Configuration::Configuration() {
 #endif
     this->umask = 0077;
     this->chroot_path = "";
+    this->full_php_process_display = false;
 }
 
 void suPHP::Configuration::readFromFile(File& file) 
@@ -160,6 +161,8 @@ void suPHP::Configuration::readFromFile(File& file)
                 this->umask = Util::octalStrToInt(value);
             else if (key == "chroot")
                 this->chroot_path = value;
+	    else if (key == "full_php_process_display")
+		this->full_php_process_display = this->strToBool(value);
             else 
                 throw ParsingException("Unknown option \"" + key + 
                                        "\" in section [global]", 
@@ -224,6 +227,10 @@ bool suPHP::Configuration::getAllowDirectoryOthersWriteable() const {
     return this->allow_directory_others_writeable;
 }
 
+bool suPHP::Configuration::getFullPHPProcessDisplay() const {
+    return this->full_php_process_display;
+}
+
 bool suPHP::Configuration::getErrorsToBrowser() const {
     return this->errors_to_browser;
 }
diff --git a/src/Configuration.hpp b/src/Configuration.hpp
index 2e99b6c..1853acb 100644
--- a/src/Configuration.hpp
+++ b/src/Configuration.hpp
@@ -59,6 +59,7 @@ namespace suPHP {
         int min_gid;
         int umask;
         std::string chroot_path;
+        bool full_php_process_display;
 
         /**
          * Converts string to bool
@@ -139,6 +140,12 @@ namespace suPHP {
         bool getAllowDirectoryOthersWriteable() const;
 
         /**
+         * Returns whether suPHP should check the target script UID in
+         * paranoid mode
+	 */
+	bool getFullPHPProcessDisplay() const;
+
+	/**
          * Returns whether (minor) error message should be sent to browser
          */
         bool getErrorsToBrowser() const;
-- 
2.2.0

